### PR TITLE
Add column-based insert interface in segcore

### PR DESCRIPTION
**Pull request by letian-jiang**
Related to issue #16085 

Signed-off-by: Letian Jiang <letian.jiang@zilliz.com>

### CODE REVIEW

File: internal/core/unittest/test_c_api.cpp [hunk 1]
```diff
@@ -216,6 +239,23 @@ TEST(CApiTest, InsertTest) {
...
+    auto collection = NewCollection(get_default_schema_config());
+    auto segment = NewSegment(collection, Growing, -1);
+
+    int N = 10000;
+    auto [raw_data, timestamps, uids] = generate_column_data(N);
+
+    int64_t offset;
+    PreInsert(segment, N, &offset);
+
+    auto res = InsertColumnData(segment, offset, N, uids.data(), timestamps.data(), raw_data.data(), N);
```
Comment (xige-16) [inline internal/core/unittest/test_c_api.cpp:252]: More tests of delete/search/query are needed

File: internal/core/src/segcore/segment_c.cpp [hunk 2]
```diff
@@ -150,6 +150,43 @@ Insert(CSegmentInterface c_segment,
...
+        milvus::segcore::ColumnBasedRawData dataChunk{};
+
+        auto& schema = segment->get_schema();
+        auto sizeof_infos = schema.get_sizeof_infos();
+        dataChunk.columns_ = std::vector<milvus::aligned_vector<uint8_t>>(schema.size());
+        // reverse space for each field
+        for (int fid = 0; fid < schema.size(); ++fid) {
+            auto len = sizeof_infos[fid];
+            dataChunk.columns_[fid].resize(len * size);
+        }
```
Comment (xige-16) [inline internal/core/src/segcore/segment_c.cpp:172]: How to insert a column of variable-length strings?

Comment (letian-jiang) [reply internal/core/src/segcore/segment_c.cpp:172]: The plan is to insert column data via an interface similar to `SegmentSealedImpl::LoadFieldData`, 
* on the Go side, pass the marshaled byte slice for each field when calling cgo
* on the C side, unmarshal to extract each field 
which will be implemented the in near future.

File: internal/core/src/segcore/segment_c.cpp [hunk 3]
```diff
@@ -150,6 +150,43 @@ Insert(CSegmentInterface c_segment,
     }
 }
 
+CStatus
+InsertColumnData(CSegmentInterface c_segment,
+                 int64_t reserved_offset,
+                 int64_t size,
+                 const int64_t* row_ids,
+                 const uint64_t* timestamps,
+                 void* raw_data,
```
Comment (czs007) [inline internal/core/src/segcore/segment_c.cpp:159]: Is it possible to pass in an array of pointers. So raw_data doesn't need to be completely flattened?

Comment (czs007) [reply internal/core/src/segcore/segment_c.cpp:159]: Columns of data from streams and batches both are multiple pointers. If you need to pass a flattened pointer, you have to copy the memory.
Am i right?

Comment (letian-jiang) [reply internal/core/src/segcore/segment_c.cpp:159]: The design generally follows the previous row-based insertion, which uses void* pointer to pass data flattened by go caller.

Comment (letian-jiang) [reply internal/core/src/segcore/segment_c.cpp:159]: I am unsure about whether the copy is redundant here. But as long as memory safety is guaranteed, we could remove copies in both insertion methods.

Comment (czs007) [reply internal/core/src/segcore/segment_c.cpp:159]: > The design generally follows the previous row-based insertion, which uses void* pointer to pass data flattened by go caller.

The previous row-based insert is about to be deprecated. I think it's unnecessary to stick to the previous design.
If go caller wants to flatten multiple contiguous memories into one contiguous memory, it has to do a memory copy. Removing this copy operation on the go side can speed up and save peak memory usage.

Comment (czs007) [reply internal/core/src/segcore/segment_c.cpp:159]: copy on the c++ side is inevitable.

Comment (letian-jiang) [reply internal/core/src/segcore/segment_c.cpp:159]: > Is it possible to pass in an array of pointers. So raw_data doesn't need to be completely flattened?

As suggested by the [cgo doc](https://pkg.go.dev/cmd/cgo), we cannot pass an array of go pointers. (Here go pointer means a pointer to memory allocated by Go)
> Go code may pass a Go pointer to C provided the Go memory to which it points does not contain any Go pointers.

Comment (letian-jiang) [reply internal/core/src/segcore/segment_c.cpp:159]: A possible solution would be something like `SegmentSealed::LoadFieldData` as mentioned by @xige-16, maybe named `SegmentGrowing::InsertFieldData`, which however requires a significant refactoring of the segcore module and is beyond the scope of this PR and my capacity as I don't quite know segcore.

Comment (xige-16) [reply internal/core/src/segcore/segment_c.cpp:159]: The original data will be marshaled according proto, and will no longer be spliced into bytes in the go code. Modifications will be in my next pr. @czs007

**Comment by xiaofan-luan** [comment]
need the following patch to remove row based interface~

**Pull request merged**